### PR TITLE
Issue 11 Estimate next purchase date

### LINKED ISSUE #11: 11. As an item, I want my estimated next purchase date to be computed at the time my purchase is recorded in the database so the app can learn how often I buy different items
## Summary
In order to advise users about when to purchase things, the app needs to be able to calculate that guess and store it a future   date.

The `calculateEstimate` function from the `@the-collab-lab/shopping-list-utils` module will help us know how many days in the future a purchase should happen. In order to use `calculateEstimate`, we need to know how many days have passed since the time the item was last purchased.

## Acceptance criteria
- [x] When the user purchases an item, the item’s `dateNextPurchased` property is calculated using the `calculateEstimate` function and saved to the Firestore database
  - [x] `dateNextPurchased` is saved as **a date**, not a number
- [x] A `getDaysBetweenDates` function is exported from `utils/dates.js`  and imported into `api/firebase.js`
  - [x] This function takes two JavaScript Dates and return the number of days that have passed between them

### Working with Firestore’s dates
**Firestore returns dates as special objects called `Timestamps.`** You will need to convert the Timestamp to a JavaScript Date in order to work with it and complete this feature. The `Timestamp` object has a `.toDate()` method – see [the Firestore docs](https://firebase.google.com/docs/reference/js/v8/firebase.firestore.Timestamp#todate).

### Calculating the time between JavaScript dates
Comparing and manipulating JavaScript Dates is notoriously tricky. Here's a hint: all Date objects have a `getTime()` method that returns a representation of that date in milliseconds ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTime)).

### Calculating the next purchase date
The `@the-collab-lab/shopping-list-utils` module should already be installed in your project. This module exports `calculateEstimate`.

`calculateEstimate` accepts three arguments. They are, in order:
- `previousEstimate` (a number): The last estimated purchase interval.
- `daysSinceLastPurchase` (a number): The number of days that have elapsed between the last purchased date and now. You will need to figure out how to calculate this argument!
- `totalPurchases` (a number): The total number of purchases made for the item,

Given this information, `calculateEstimate` returns **a whole number** indicating the number of days until the user is likely to make their next purchase.

### PR DESCRIPTION
**Pull request by NickRoccodev11**
## Description
This ticket creates an estimate for the next purchase date of an item. 
- It bases the estimate on the users purchasing history, comparing the intervals between purchase dates while considering the number of total purchases.
- Working on this ticket helped improve our logic for calculating future dates.
- We ran into issues regarding copying Dates, as they are references.

## Related Issue
closes #11

## Acceptance Criteria
 

- [ ] When the user purchases an item, the item’s dateNextPurchased property is calculated using the calculateEstimate function and saved to the Firestore database

- [ ] dateNextPurchased is saved as a date, not a number
- [ ] A getDaysBetweenDates function is exported from utils/dates.js and imported into api/firebase.js
 - [ ] This function takes two JavaScript Dates and return the number of days that have passed between them

### Before
issue11before

### After
issue11after

## Testing Steps / QA Criteria
- Open the Firestore database.
- Locate an item in your list that you want to purchase, and make note of `dateNextPurchased ` date.
- Sign In to app using email account.
- Click on a list and navigate to `List ` view.
- Purchase the item you previously observed in Firestore.
- After successful purchase, return to Firestore to view the item.
- Confirm that `dateNextPurchased ` has changed to a future date. 
- If you are interested in seeing how the estimate is calculated given various conditions, refer to this documentation: https://github.com/the-collab-lab/shopping-list-utils/blob/main/src/calculateEstimate/calculateEstimate.ts

### CODE REVIEW

**Review by dterceroparker — APPROVED** [review]
** 🦋 praise:** I tested two separate lists and six items, and not only is `dateLastPurchased` no longer null, but `dateNextPurchased` is accurately updating based on `dateLastPurchased` and `purchaseInterval`. The added documentation in the PR was very helpful.

Well done 🎊! Your work on this feature would make a great LinkedIn post 😉

**Review by shuveksha-tuladhar — APPROVED** [review]
@NickRoccodev11 and @stacy-tech Great work! I am approving this code.

But, I have a question about how the `purchaseInterval` is being updated. I added a new item with a `purchaseInterval` of 14 and then I purchased that item from the List page. 

When I checked data on the Firebase for that particular item, the `purchaseInterval` was reset to 0. 

Is this the expected behavior for the `purchaseInterval`, or should it retain the previous value of 14 for that item? 

Here is the workflow of how I tested it:

https://github.com/user-attachments/assets/21c206ec-add6-4905-9843-d694bda88443

**Comment by NickRoccodev11** [comment]
@shuveksha-tuladhar Thanks for the video explanation! so yes, for now that is the expected behavior. the ` calculateEstimate ` function first checks the total number of purchases an item has. if it has been purchased less than twice, the function returns `daysSinceLastPurchase` . (or "daysSinceLastTransaction" as it's labeled in the documentation linked above)  What this means is, if you create an item and buy it on the same day, 0 days have passed since the item was last interacted with, so it sets the interval to 0. This might not be the behavior we ultimately want (considering adding an item to the list isn't technically purchasing it) but it is what we expect at this point

File: src/utils/dates.js [hunk 1]
```diff
@@ -10,3 +10,9 @@ const ONE_DAY_IN_MILLISECONDS = 86400000;
 export function getFutureDate(offset) {
 	return new Date(Date.now() + offset * ONE_DAY_IN_MILLISECONDS);
 }
+export function getDaysBetweenDates(previousPurchaseDate) {
```
Comment (jtkabenni) [inline src/utils/dates.js:13]: i would consider what to name this function - if i wanted to get the days between dates, i would expect this function to accept two values, date1 and date 2. Since you are only passing in one date and compare it to today, i would name this function something to reflect that. 

This is so when this util function is needed for something else in the future it's easy to tell what it can do.

File: src/utils/dates.js [hunk 2]
```diff
@@ -10,3 +10,9 @@ const ONE_DAY_IN_MILLISECONDS = 86400000;
 export function getFutureDate(offset) {
```
Comment (jtkabenni) [inline src/utils/dates.js:10]: I would also reconsider making this function name more specific

File: src/api/firebase.js [hunk 3]
```diff
@@ -195,23 +195,44 @@ export async function addItem(listPath, { itemName, daysUntilNextPurchase }) {
...
+	const daysUntilNextPurchase = calculateEstimate(
+		purchaseInterval,
+		daysSinceLastPurchase,
 		totalPurchases,
-		dateLastPurchased,
+	);
+
+	purchaseInterval = daysUntilNextPurchase;
+
+	const dateNextPurchased = getFutureDate(daysUntilNextPurchase);
```
Comment (jtkabenni) [inline src/api/firebase.js]: I think your logic makes sense here, however it is a little bit hard to follow (in nature due to the calculation itself being tricky. I would consider adding comments to each line to make sure it is clear what each line is doing. 

This also feels like quite a big block of code. I would consider trying to separate out anything that is not the purpose of the `updateItem` function  (to update the database with the item's updated data), and seeing where else I could put this calculating dates logic - maybe in the list component itself, or another util function? You will then need to change the arguments for this function. This way we can keep this main function less noisy.

**Review by jtkabenni — APPROVED** [review]
Looks good and the behavior is working as expected! See the notes I added. Nothing is a blocker though.

File: src/api/firebase.js (hunk 3 above)
Comment (stacy-tech) [reply src/api/firebase.js]: Hi Jenny,
Thank you for your feedback! I appreciate your suggestion about adding comments to clarify the code. I’ve added comments to each line to explain what’s happening. This should make the logic more transparent.
